{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getTrID(idpG4):\n",
    "    return( idpG4.split('|')[0] )\n",
    "\n",
    "def getChr(idpG4):\n",
    "    return( idpG4.split('|')[1].split(':')[0] )\n",
    "\n",
    "def getpG4Start(idpG4):\n",
    "    return( idpG4.split('|')[1].split(':')[1].split('-')[0] )\n",
    "\n",
    "def getpG4End(idpG4):\n",
    "    return( idpG4.split('|')[1].split(':')[1].split('-')[0] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/anais/.local/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Strand</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>locStart</th>\n",
       "      <th>locEnd</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Location</th>\n",
       "      <th>meancGcC</th>\n",
       "      <th>meanG4H</th>\n",
       "      <th>meanG4NN</th>\n",
       "      <th>pG4Start</th>\n",
       "      <th>pG4End</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Biotype</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Class</th>\n",
       "      <th>Length</th>\n",
       "      <th>LengthTr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>18773374</td>\n",
       "      <td>18790569</td>\n",
       "      <td>ENSG00000176728</td>\n",
       "      <td>intron</td>\n",
       "      <td>8.571429</td>\n",
       "      <td>1.116667</td>\n",
       "      <td>0.709113</td>\n",
       "      <td>18789497</td>\n",
       "      <td>18789438</td>\n",
       "      <td>ENST00000651886</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>CCTGGTCAAAAGGAGGGCTCTATTCTGATGCCTGTGGTGGGGGCGG...</td>\n",
       "      <td>LongNC</td>\n",
       "      <td>59</td>\n",
       "      <td>17195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>18773374</td>\n",
       "      <td>18790569</td>\n",
       "      <td>ENSG00000176728</td>\n",
       "      <td>intron</td>\n",
       "      <td>67.000000</td>\n",
       "      <td>0.983333</td>\n",
       "      <td>0.619213</td>\n",
       "      <td>18786537</td>\n",
       "      <td>18786478</td>\n",
       "      <td>ENST00000651886</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TGAAGGGTGATATGGTAATACTAAGGATGGGAGAGAAGTTTGGGAT...</td>\n",
       "      <td>LongNC</td>\n",
       "      <td>59</td>\n",
       "      <td>17195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>18773374</td>\n",
       "      <td>18790569</td>\n",
       "      <td>ENSG00000176728</td>\n",
       "      <td>intron</td>\n",
       "      <td>21.166667</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.854671</td>\n",
       "      <td>18786457</td>\n",
       "      <td>18786388</td>\n",
       "      <td>ENST00000651886</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TAGATTGAGGGTTAGGTCTGGGGTTGACTTAAAGGTTCGTGTGGTG...</td>\n",
       "      <td>LongNC</td>\n",
       "      <td>69</td>\n",
       "      <td>17195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>18773374</td>\n",
       "      <td>18790569</td>\n",
       "      <td>ENSG00000176728</td>\n",
       "      <td>intron</td>\n",
       "      <td>16.166667</td>\n",
       "      <td>0.933333</td>\n",
       "      <td>0.823813</td>\n",
       "      <td>18786287</td>\n",
       "      <td>18786208</td>\n",
       "      <td>ENST00000651886</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>GGTAAGAGATAGGCTTACAGTTAGATGTTGGGGTTAGTGGAGGGTT...</td>\n",
       "      <td>LongNC</td>\n",
       "      <td>79</td>\n",
       "      <td>17195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-</td>\n",
       "      <td>Y</td>\n",
       "      <td>18773374</td>\n",
       "      <td>18790569</td>\n",
       "      <td>ENSG00000176728</td>\n",
       "      <td>intron</td>\n",
       "      <td>41.433333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.824185</td>\n",
       "      <td>18786017</td>\n",
       "      <td>18785848</td>\n",
       "      <td>ENST00000651886</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TAGGGTTATTGTTTAGCACTTAGGTTTAGGTTTAGGGTTGGTTGAG...</td>\n",
       "      <td>LongNC</td>\n",
       "      <td>169</td>\n",
       "      <td>17195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Strand Chromosome  locStart    locEnd             Gene Location   meancGcC  \\\n",
       "0      -          Y  18773374  18790569  ENSG00000176728   intron   8.571429   \n",
       "1      -          Y  18773374  18790569  ENSG00000176728   intron  67.000000   \n",
       "2      -          Y  18773374  18790569  ENSG00000176728   intron  21.166667   \n",
       "3      -          Y  18773374  18790569  ENSG00000176728   intron  16.166667   \n",
       "4      -          Y  18773374  18790569  ENSG00000176728   intron  41.433333   \n",
       "\n",
       "    meanG4H  meanG4NN  pG4Start    pG4End       Transcript Biotype  \\\n",
       "0  1.116667  0.709113  18789497  18789438  ENST00000651886  lncRNA   \n",
       "1  0.983333  0.619213  18786537  18786478  ENST00000651886  lncRNA   \n",
       "2  0.933333  0.854671  18786457  18786388  ENST00000651886  lncRNA   \n",
       "3  0.933333  0.823813  18786287  18786208  ENST00000651886  lncRNA   \n",
       "4  1.000000  0.824185  18786017  18785848  ENST00000651886  lncRNA   \n",
       "\n",
       "                                            Sequence   Class  Length  LengthTr  \n",
       "0  CCTGGTCAAAAGGAGGGCTCTATTCTGATGCCTGTGGTGGGGGCGG...  LongNC      59     17195  \n",
       "1  TGAAGGGTGATATGGTAATACTAAGGATGGGAGAGAAGTTTGGGAT...  LongNC      59     17195  \n",
       "2  TAGATTGAGGGTTAGGTCTGGGGTTGACTTAAAGGTTCGTGTGGTG...  LongNC      69     17195  \n",
       "3  GGTAAGAGATAGGCTTACAGTTAGATGTTGGGGTTAGTGGAGGGTT...  LongNC      79     17195  \n",
       "4  TAGGGTTATTGTTTAGCACTTAGGTTTAGGTTTAGGGTTGGTTGAG...  LongNC     169     17195  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfWTMoi = pd.read_csv('/home/anais/Documents/Projet/G4Conservation/data/homo_sapiens/pG4WT.csv', sep='\\t')\n",
    "dfWTMoi['Length'] = dfWTMoi['pG4End'] - dfWTMoi['pG4Start']\n",
    "dfWTMoi['Length'] = abs( dfWTMoi['Length'] )\n",
    "dfWTMoi['LengthTr'] = dfWTMoi['locEnd'] - dfWTMoi['locStart']\n",
    "dfWTMoi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>InfoG4ByTranscript</th>\n",
       "      <th>cGcC</th>\n",
       "      <th>G4Hunter</th>\n",
       "      <th>sequenceG4</th>\n",
       "      <th>G4NN</th>\n",
       "      <th>location</th>\n",
       "      <th>transcriptBiotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000638249|10:77289210-77289269|-1</td>\n",
       "      <td>10.875</td>\n",
       "      <td>1.13333333333</td>\n",
       "      <td>TTGGGGCTGGCTGATGCTGGGGAGTGGGACTGTCTGCTGGGTTCAA...</td>\n",
       "      <td>0.91178512915</td>\n",
       "      <td>Intron</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000299366|10:71629784-71629853|1</td>\n",
       "      <td>11.7142857143</td>\n",
       "      <td>0.983333333333</td>\n",
       "      <td>CACGGGAAGCAGATTAGTGGTTGCCAAGGGTTGGGGGAGAGAAATG...</td>\n",
       "      <td>0.757842542274</td>\n",
       "      <td>Intron</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000465182|10:101354913-101354972|1</td>\n",
       "      <td>8.44444444444</td>\n",
       "      <td>0.9</td>\n",
       "      <td>CCTGGGAGGGGAAAAGGGTGTTGACTGTAAGCTAGAATTACAAATG...</td>\n",
       "      <td>0.755268599579</td>\n",
       "      <td>IntronNC</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000419399|10:49750470-49750529|-1</td>\n",
       "      <td>6.33333333333</td>\n",
       "      <td>0.983333333333</td>\n",
       "      <td>GACAAGGAGAGGTGGAGGGATTTCTTCCAGGGAAGGGAGGGCAGGG...</td>\n",
       "      <td>0.603612229916</td>\n",
       "      <td>Intron</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000640605|10:77529190-77529279|-1</td>\n",
       "      <td>13.6263888889</td>\n",
       "      <td>1.19166666666</td>\n",
       "      <td>GAGGTGAGCTGGGAAACCACAGGCAGAGTGGTGCAGGGGGAAAAAG...</td>\n",
       "      <td>0.857757786925</td>\n",
       "      <td>Intron</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         InfoG4ByTranscript           cGcC        G4Hunter  \\\n",
       "0   ENST00000638249|10:77289210-77289269|-1         10.875   1.13333333333   \n",
       "1    ENST00000299366|10:71629784-71629853|1  11.7142857143  0.983333333333   \n",
       "2  ENST00000465182|10:101354913-101354972|1  8.44444444444             0.9   \n",
       "3   ENST00000419399|10:49750470-49750529|-1  6.33333333333  0.983333333333   \n",
       "4   ENST00000640605|10:77529190-77529279|-1  13.6263888889   1.19166666666   \n",
       "\n",
       "                                          sequenceG4            G4NN  \\\n",
       "0  TTGGGGCTGGCTGATGCTGGGGAGTGGGACTGTCTGCTGGGTTCAA...   0.91178512915   \n",
       "1  CACGGGAAGCAGATTAGTGGTTGCCAAGGGTTGGGGGAGAGAAATG...  0.757842542274   \n",
       "2  CCTGGGAGGGGAAAAGGGTGTTGACTGTAAGCTAGAATTACAAATG...  0.755268599579   \n",
       "3  GACAAGGAGAGGTGGAGGGATTTCTTCCAGGGAAGGGAGGGCAGGG...  0.603612229916   \n",
       "4  GAGGTGAGCTGGGAAACCACAGGCAGAGTGGTGCAGGGGGAAAAAG...  0.857757786925   \n",
       "\n",
       "   location        transcriptBiotype  \n",
       "0    Intron  nonsense_mediated_decay  \n",
       "1    Intron           protein_coding  \n",
       "2  IntronNC     processed_transcript  \n",
       "3    Intron           protein_coding  \n",
       "4    Intron           protein_coding  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfSarah = pd.read_csv('/home/anais/Documents/Projet/G4Conservation/data/homo_sapiens/HS_All_G4InTranscript.txt', sep='\\t')\n",
    "dfSarah['Transcript'] = dfSarah.InfoG4ByTranscript.apply(getTrID)\n",
    "dfSarah.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
